### PR TITLE
Trajectory Validation (and fixing)

### PR DESCRIPTION
**Pull request by henningkayser**
This effort adds a check to the execute() function to validate all waypoints are reachable within the given time stamps and all values are within bounds. In case a trajectory is invalid an InvalidTrajectoryExecution is thrown. An optional parameter fix_trajectory should allow reparameterizing invalid trajectories.

### CODE REVIEW

**Comment by henningkayser** [comment]
This version contains the copied sources for TOTG since currently moveit_simple only targets kinetic where TOTG has not been merged.

File: moveit_simple/include/moveit/trajectory_processing/time_optimal_trajectory_generation.h [hunk 1]
```diff
@@ -0,0 +1,178 @@
+/*
```
Comment (davetcoleman) [inline moveit_simple/include/moveit/trajectory_processing/time_optimal_trajectory_generation.h]: I really don't like this code duplication

Comment (henningkayser) [reply moveit_simple/include/moveit/trajectory_processing/time_optimal_trajectory_generation.h]: Me neither. I think it's more like a confirmation that it works as intended. I would suggest pushing moveit_simple to melodic first and then creating a new PR without these.

File: moveit_simple/CMakeLists.txt [hunk 2]
```diff
@@ -68,8 +72,10 @@ add_library(${PROJECT_NAME}
 add_dependencies(${PROJECT_NAME} ${catkin_EXPORTED_TARGETS})
 add_dependencies(${PROJECT_NAME} ${PROJECT_NAME}_gencfg)
 add_dependencies(${PROJECT_NAME} ${PROJECT_NAME}_genparam)
+add_dependencies(${PROJECT_NAME} moveit_core_time_optimal_trajectory_generation)
 
 target_link_libraries(${PROJECT_NAME} ${catkin_LIBRARIES})
+target_link_libraries(${PROJECT_NAME} moveit_core_time_optimal_trajectory_generation)
```
Comment (mlautman) [inline moveit_simple/CMakeLists.txt]: combine with previous line

File: moveit_simple/CMakeLists.txt [hunk 3]
```diff
@@ -57,6 +57,10 @@ include_directories(
   ${EIGEN3_INCLUDE_DIRS}
 )
 
+add_library(moveit_core_time_optimal_trajectory_generation
```
Comment (mlautman) [inline moveit_simple/CMakeLists.txt]: I would prefix this with `${PROJECT_NAME}` (eg: `${PROJECT_NAME}_time_optimal_trajectory_generation`

File: moveit_simple/include/moveit/trajectory_processing/time_optimal_trajectory_generation.h (hunk 1 above)
Comment (mlautman) [reply moveit_simple/include/moveit/trajectory_processing/time_optimal_trajectory_generation.h]: Add a TODO: to remove this when we move to Melodic

File: moveit_simple/include/moveit_simple/online_robot.h [hunk 4]
```diff
@@ -59,23 +59,26 @@ class OnlineRobot : public Robot
    * @param traj_name - name of trajectory to be executed (must be filled with
    * prior calls to "addTrajPoint".
    * @param collision_check - bool to turn check for collision on\off
+   * @param fix_trajectory - bool to fix trajectory timesteps if necessary on\off
```
Comment (mlautman) [inline moveit_simple/include/moveit_simple/online_robot.h]: maybe a nit but I don't like `fix_trajectory` as it is too general. How about `retime_trajectory`

File: moveit_simple/include/moveit_simple/online_robot.h [hunk 5]
```diff
@@ -59,23 +59,26 @@ class OnlineRobot : public Robot
...
    * in Collision with itself or environment)
    */
-  void execute(const std::string traj_name, bool collision_check = false);
+  void execute(const std::string traj_name, bool collision_check = false, bool fix_trajectory = false);
 
   /**
    * @brief execute a given planned joint trajectory
    * @param goal - Joint Trajectory goal which is a known 'Plan'
    * @param collision_check - bool to turn check for collision on\off
+   * @param fix_trajectory - bool to fix trajectory timesteps if necessary on\off
```
Comment (mlautman) [inline moveit_simple/include/moveit_simple/online_robot.h]: same

File: moveit_simple/test/motoman_mh5.cpp [hunk 6]
```diff
@@ -746,18 +766,18 @@ TEST_F(UserRobotTest, copy_current_pose)
 
   // Before the first execution robot should be at position "home"
   before_execution_1 = user_robot->getJointState();
-  EXPECT_NO_THROW(user_robot->execute(TRAJECTORY_NAME));
+  EXPECT_NO_THROW(user_robot->execute(TRAJECTORY_NAME, false, true));
```
Comment (mlautman) [inline moveit_simple/test/motoman_mh5.cpp:770]: create `bool check_collisions = true` and `bool fix_trajectory = false` and use them here to make the code more readable. the compiler should optimize it out anyway

File: moveit_simple/include/moveit_simple/trajectory_processing.h [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        }
+      }
+    }
+    previous_point = point;
+  }
+  result.value = TrajectoryValidationResult::Success;
+  return result;
+}
+
+void fixTrajectory(robot_model::RobotModelConstPtr robot_model, const std::string& group_name,
```
Comment (mlautman) [inline moveit_simple/include/moveit_simple/trajectory_processing.h]: rename to `retimeTrajectory`

File: moveit_simple/include/moveit_simple/trajectory_processing.h [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+         InvalidAcceleration,
+         InvalidTimestamp} value;
+  std::string error_message;
+};
+TrajectoryValidationResult validateTrajectory(robot_model::RobotModelConstPtr robot_model,
+                                              const trajectory_msgs::JointTrajectory& trajectory)
+{
+  TrajectoryValidationResult result;
+  const std::vector<std::string>& joint_names = trajectory.joint_names;
+  trajectory_msgs::JointTrajectoryPoint previous_point;
```
Comment (mlautman) [inline moveit_simple/include/moveit_simple/trajectory_processing.h:44]: Where is this set for the first point?

File: moveit_simple/include/moveit_simple/trajectory_processing.h [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    const auto& point = trajectory.points[i];
+    for (size_t j = 0; j < joint_names.size(); ++j)
+    {
+      const auto& joint_model = robot_model->getJointModel(joint_names[j]);
+      const auto& bounds = joint_model->getVariableBounds()[0];
+      // validate position bounds and throw exception since this can't be fixed easily
+      if (bounds.position_bounded_ && !joint_model->satisfiesPositionBounds(&point.positions[j]))
+      {
+        result.value = TrajectoryValidationResult::InvalidPosition;
+        result.error_message = "Trajectory contains waypoints that don't satisfy position bounds";
```
Comment (mlautman) [inline moveit_simple/include/moveit_simple/trajectory_processing.h]: I would add more diagnostic information (eg. which joint, which index, what the invalid value was and what the bound was)

File: moveit_simple/include/moveit_simple/trajectory_processing.h [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        return result;
+      }
+      // validate velocity/acceleration bounds and pair-wise time difference
+      double dt = (point.time_from_start - previous_point.time_from_start).toSec();
+      if (bounds.velocity_bounded_)
+      {
+        result.value = TrajectoryValidationResult::InvalidVelocity;
```
Comment (mlautman) [inline moveit_simple/include/moveit_simple/trajectory_processing.h]: Shouldn't you wait until after the velocity check to set the result.value to an error?

Comment (mlautman) [reply moveit_simple/include/moveit_simple/trajectory_processing.h]: Ahhh, I see that this works. Maybe just a nit-pick then. It's easier to follow the logic if this is only set when the value is invalid

File: moveit_simple/test/kuka_kr210.cpp [hunk 11]
```diff
@@ -146,6 +146,25 @@ TEST_F(UserRobotTest, add_trajectory)
...
+TEST_F(UserRobotTest, validate_trajectory)
+{
+  ROS_INFO_STREAM("Testing trajectory validation");
+  const std::string TRAJECTORY_NAME("val_traj");
+  const moveit_simple::InterpolationType joint = moveit_simple::interpolation_type::JOINT;
+
+  // try valid trajectory
+  EXPECT_NO_THROW(robot->addTrajPoint(TRAJECTORY_NAME, "home",      0.5));
+  EXPECT_NO_THROW(robot->addTrajPoint(TRAJECTORY_NAME, "waypoint1", 1.0, joint, 5));
+  EXPECT_NO_THROW(robot->execute(TRAJECTORY_NAME));
```
Comment (mlautman) [inline moveit_simple/test/kuka_kr210.cpp:158]: Suggestion: Add a test that calls `trajectory_processing::validateTrajectory` directly?

File: moveit_simple/include/moveit_simple/trajectory_processing.h (hunk 8 above)
Comment (mlautman) [reply moveit_simple/include/moveit_simple/trajectory_processing.h:44]: Ahh I see that you only reference if `i>0`

**Comment by henningkayser** [comment]
@davetcoleman @mlautman I'll have this rebased onto #4 and address the feedback first thing next week.

**Comment by mlautman** [comment]
Let's point this at the upstream repo with a note that it should target a devel branch instead

File: moveit_simple/include/moveit/trajectory_processing/time_optimal_trajectory_generation.h (hunk 1 above)
Comment (davetcoleman) [reply moveit_simple/include/moveit/trajectory_processing/time_optimal_trajectory_generation.h]: lets just move to master branch now

**Comment by davetcoleman** [comment]
great feedback mike!

**Pull request closed without merging**